pnpm install sometimes never exits after printing `Done in Xs` (CI, Linux, pnpm 11.x)

### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

11.5.1 (also present across 11.x; just upgraded to 11.5.2 — nothing in its changelog looked related, will report back if it fixes it)

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_Not deterministically reproducible — happens on ~5% of CI runs in a private monorepo (40 workspace packages)._

### Reproduction steps

1. GitHub Actions self-hosted runner, Linux x64, Node 24.12.0, pnpm 11.5.1 via corepack
2. pnpm workspace, 40 packages, `nodeLinker: hoisted`, `supportedArchitectures` restricted to `current`
3. Restore a partial `node_modules` from CI cache (root `node_modules` only, so install re-fetches ~330 tarballs)
4. Run `pnpm install --frozen-lockfile`

### Describe the Bug

On roughly 5% of CI runs (9 of our last 200), `pnpm install` completes successfully — postinstall runs, `Done in Xs` is printed — but the pnpm process **never exits**. Nothing more is logged; the GitHub Actions job sits there until its 30-minute timeout kills it, and the runner cleanup reports a leftover orphan process.

```
2026-06-09T16:37:37.626Z [WARN] GET https://registry.npmjs.org/@rolldown/binding-linux-ppc64-gnu/-/binding-linux-ppc64-gnu-1.0.3.tgz error (ECONNRESET). Will retry in 10 seconds. 2 retries left.
2026-06-09T16:37:43.049Z Progress: resolved 0, reused 0, downloaded 325, added 34, done
2026-06-09T16:37:44.030Z . postinstall$ husky
2026-06-09T16:37:44.167Z . postinstall: Done
2026-06-09T16:37:44.338Z Done in 14.6s using pnpm v11.5.1
2026-06-09T17:07:21.329Z ##[error]The operation was canceled.   <- job timeout 30 min later
2026-06-09T17:07:21.618Z Terminate orphan process: pid (590) (MainThread)
```

Observations across the hung runs:

- The install itself always succeeded (`Done in ~15s`, node_modules complete — a retry of the job is a fast no-op install).
- All hung runs had performed real network fetches (~330 tarballs). Some — not all — show `ECONNRESET ... Will retry` warnings shortly before completion, so lingering retry/keep-alive sockets are a plausible suspect for keeping the event loop alive.
- Runs whose install is a no-op (`Already up to date`) have never hung for us.

Same symptom as #6964 (pnpm 8, reportedly fixed in 8.7.3), reopened territory on 11.x apparently.

### Expected Behavior

The pnpm process exits once `Done in Xs` is printed.

### Which Node.js version are you using?

24.12.0

### Which operating systems have you used?

- [X] Linux
- [ ] macOS
- [ ] Windows

### If your OS is a Linux based, which one it is? (leave empty if not applicable)

Ubuntu (GitHub Actions self-hosted runner image)

--- a/.changeset/keep-worker-pool-for-stragglers.md
+++ b/.changeset/keep-worker-pool-for-stragglers.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/worker": patch
+"pnpm": patch
+---
+
+Fixed `pnpm install` sometimes never exiting after printing `Done in Xs` [#12297](https://github.com/pnpm/pnpm/issues/12297). A worker call arriving after the CLI's final `finishWorkers()` — typically a tarball fetch for a package that is never linked (e.g. a foreign-architecture optional dependency), delayed past install completion by a network retry — lazily re-created the worker pool, and nothing ever finished the new pool, so its idle worker thread kept the process alive until an external timeout killed it. `finishWorkers()` now keeps the finished pool's reference: straggler tasks are still serviced, and their workers are torn down again at check-in.

--- a/.changeset/keep-worker-pool-for-stragglers.md
+++ b/.changeset/keep-worker-pool-for-stragglers.md
@@ -3,4 +3,4 @@
 "pnpm": patch
 ---
 
-Fixed `pnpm install` sometimes never exiting after printing `Done in Xs` [#12297](https://github.com/pnpm/pnpm/issues/12297). A worker call arriving after the CLI's final `finishWorkers()` — typically a tarball fetch for a package that is never linked (e.g. a foreign-architecture optional dependency), delayed past install completion by a network retry — lazily re-created the worker pool, and nothing ever finished the new pool, so its idle worker thread kept the process alive until an external timeout killed it. `finishWorkers()` now keeps the finished pool's reference: straggler tasks are still serviced, and their workers are torn down again at check-in.
+Fixed `pnpm install` sometimes not exiting after printing `Done in Xs` [#12297](https://github.com/pnpm/pnpm/issues/12297).

--- a/pnpm11/worker/src/index.ts
+++ b/pnpm11/worker/src/index.ts
@@ -33,7 +33,13 @@ export async function finishWorkers (): Promise<void> {
   // @ts-expect-error
   global.finishWorkers = undefined
   await finish?.()
-  workerPool = undefined
+  // Deliberately keep the `workerPool` reference. The finished pool still
+  // services straggler tasks (e.g. a tarball fetch delayed by a network retry
+  // that completes after the CLI's final drain): each such task spawns a
+  // worker that is torn down again at check-in because the pool is finishing.
+  // Clearing the reference would make stragglers lazily create a fresh pool
+  // that nothing ever finishes, leaving an idle worker thread that keeps the
+  // process alive forever (#12297).
 }
 
 function createTarballWorkerPool (): WorkerPool {

--- a/pnpm11/worker/src/index.ts
+++ b/pnpm11/worker/src/index.ts
@@ -33,13 +33,6 @@ export async function finishWorkers (): Promise<void> {
   // @ts-expect-error
   global.finishWorkers = undefined
   await finish?.()
-  // Deliberately keep the `workerPool` reference. The finished pool still
-  // services straggler tasks (e.g. a tarball fetch delayed by a network retry
-  // that completes after the CLI's final drain): each such task spawns a
-  // worker that is torn down again at check-in because the pool is finishing.
-  // Clearing the reference would make stragglers lazily create a fresh pool
-  // that nothing ever finishes, leaving an idle worker thread that keeps the
-  // process alive forever (#12297).
 }
 
 function createTarballWorkerPool (): WorkerPool {

--- a/pnpm11/worker/test/finishWorkers.test.ts
+++ b/pnpm11/worker/test/finishWorkers.test.ts
@@ -1,0 +1,39 @@
+import { spawn } from 'node:child_process'
+import fs from 'node:fs'
+import os from 'node:os'
+import path from 'node:path'
+
+import { expect, test } from '@jest/globals'
+
+const libIndex = new URL('../lib/index.js', import.meta.url).href
+
+// Regression test for https://github.com/pnpm/pnpm/issues/12297: a worker
+// call that arrives after finishWorkers() (e.g. a tarball fetch delayed by a
+// network retry, finishing after the CLI's final drain) must not leave a
+// worker thread alive — a leaked idle worker keeps the process from exiting.
+test('process exits after a worker call that arrives after finishWorkers()', async () => {
+  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'pnpm-worker-straggler-'))
+  const storeDir = path.join(tmp, 'store')
+  fs.mkdirSync(storeDir)
+  const scenario = path.join(tmp, 'scenario.mjs')
+  fs.writeFileSync(scenario, `
+const { finishWorkers, readPkgFromCafs } = await import(${JSON.stringify(libIndex)})
+const ctx = { storeDir: ${JSON.stringify(storeDir)}, verifyStoreIntegrity: false }
+await readPkgFromCafs(ctx, ${JSON.stringify(path.join(storeDir, 'idx-1'))}) // normal install work
+await finishWorkers() // the CLI's final drain
+await readPkgFromCafs(ctx, ${JSON.stringify(path.join(storeDir, 'idx-2'))}) // late straggler task
+`)
+  const exitCode = await new Promise((resolve, reject) => {
+    const child = spawn(process.execPath, [scenario], { stdio: 'ignore' })
+    const timer = setTimeout(() => {
+      child.kill('SIGKILL')
+      resolve('leaked worker kept the process alive')
+    }, 8000)
+    child.on('error', reject)
+    child.on('exit', (code) => {
+      clearTimeout(timer)
+      resolve(code)
+    })
+  })
+  expect(exitCode).toBe(0)
+}, 15000)

